src/SIL.Harmony/CrdtConfig.cs

```diff
@@ -103,52 +105,66 @@ public class ObjectTypeListBuilder
     /// </summary>
     public void Freeze()
     {
+        if (_frozen) return;
         _frozen = true;
+        JsonTypes = AdapterProvider.JsonTypes;
+        foreach (var registration in AdapterProvider.GetRegistrations())
+        {
+            ModelConfigurations.Add((builder, config) =>
+            {
+                if (!config.EnableProjectedTables) return;
+                var entity = registration.EntityBuilder(builder);
+                entity.HasOne(typeof(ObjectSnapshot))
+                    .WithOne()
+                    .HasForeignKey(registration.ObjectDbType, ObjectSnapshot.ShadowRefName)
+                    .OnDelete(DeleteBehavior.SetNull);
+            });
+        }
     }
 
-    private void CheckFrozen()
+    internal void CheckFrozen()
     {
         if (_frozen) throw new InvalidOperationException($"{nameof(ObjectTypeListBuilder)} is frozen");
     }
 
-    internal List<JsonDerivedType> Types { get; } = [];
+    internal Dictionary<Type, List<JsonDerivedType>>? JsonTypes { get; set; }
 
     internal List<Action<ModelBuilder, CrdtConfig>> ModelConfigurations { get; } = [];
 
-    public ObjectTypeListBuilder AddDbModelConfig(Action<ModelBuilder> modelConfiguration)
+    internal IObjectAdapterProvider AdapterProvider => _adapterProvider ?? throw new InvalidOperationException("No adapter has been added to the builder");
+    private IObjectAdapterProvider? _adapterProvider;
+
+    public DefaultAdapterProvider DefaultAdapter()
     {
         CheckFrozen();
-        ModelConfigurations.Add((builder, _) => modelConfiguration(builder));
-        return this;
+        if (_adapterProvider is not null) throw new InvalidOperationException("adapter has already been added");
+        var adapter = new DefaultAdapterProvider(this);
+        _adapterProvider = adapter;
+        return adapter;
     }
-
-  
-    public ObjectTypeListBuilder Add<TDerived>(Action<EntityTypeBuilder<TDerived>>? configureDb = null)
-        where TDerived : class, IObjectBase
+    
+    /// <summary>
+    /// add a custom adapter for a common interface
+    /// this is required as CRDT objects must express their references and have an Id property
+    /// using a custom adapter allows your model to not take a dependency on Harmony
+    /// </summary>
+    /// <typeparam name="TCommonInterface">
+    /// A common interface that all objects in your application implement
+    /// which System.Text.Json will deserialize your objects to, they must support polymorphic deserialization
+    /// </typeparam>
+    /// <typeparam name="TAdapter">
+    /// This adapter will be serialized and stored in the database,
```

Comment:
it's complicated. The adapter is stored in the snapshot because otherwise it couldn't be JSON deserialized. However the actual object IS stored in the projected tables as those don't have the same JSON issues